### PR TITLE
Detailed design: cancellation and nurseries 

### PR DESCRIPTION
**Pull request by ktoso**
Work in progress... not adding new things semantics so far

### CODE REVIEW

File: proposals/nnnn-structured-concurrency.md [hunk 1]
```diff
@@ -255,6 +257,148 @@ If the scope of an `async let` exits with a thrown error, the child task corresp
...
+Let us break up the `chopOnionsAndCarrots()` function into multiple steps to fully understand its semantics:
+
+- first w add vegetable chopping tasks to the nursery
+- the chopping of the various vegetables beings asynchronously,
+- eventually an onion will be chopped and `throw`
+
+
+So far we did not yet discuss the cancellation of nurseries. A nursery can be cancelled form the outside, for example like this: 
+
+#### Nurseries: Implicitly awaited tasks
```
Comment (ktoso) [inline proposals/nnnn-structured-concurrency.md:501]: Could you have a look @rjmccall if this is the intended semantics?

This is how things work in kotlin and kind of how I always thought about it (tbh even without doing the explicit add dance).

File: proposals/nnnn-structured-concurrency.md [hunk 2]
```diff
@@ -75,9 +75,11 @@ Bringing it back to our example, note that the `chopVegetables()` function might
 
 The `async let` construct makes it easy to create a set number of child tasks and associate them with variables. However, the construct does not work as well with dynamic workloads, where we don't know the number child tasks we will need to create because (for example) it is dependent on the size of a data structure. For that, we need a more dynamic construct: a task *nursery*.
 
-A nursery defines a scope in which one can create new child tasks programmatically. As with all child tasks, the child tasks within the nursery must complete when the scope exits or they will be implicitly cancelled. Nurseries also provide utilities for working with the child tasks, e.g., by waiting until the next child task completes.
+A nursery defines a scope in which one can create new child tasks programmatically. As with all child tasks, the child tasks within the nursery must complete when the scope ends or they will be implicitly cancelled. Nurseries also provide utilities for working with the child tasks, e.g., by waiting until the next child task completes.
```
Comment (ktoso) [inline proposals/nnnn-structured-concurrency.md]: or is "scope exits" specific wording that we always use and should stick to it?

Comment (rjmccall) [reply proposals/nnnn-structured-concurrency.md]: It's somewhat conventional, but it's not a big deal.

Comment (ktoso) [reply proposals/nnnn-structured-concurrency.md]: Ah i see, will return it back to exists then

File: proposals/nnnn-structured-concurrency.md [hunk 3]
```diff
@@ -255,6 +257,148 @@ If the scope of an `async let` exits with a thrown error, the child task corresp
...
+  // could be propagated out of the callback body reasonably, unless we
+  // commit to doing multi-statement closure typechecking.
+  public static func withNursery<TaskResult, BodyResult>(
+    resultType: TaskResult.Type,          
+    body: (inout Nursery<TaskResult>) async throws -> BodyResult
+  ) async rethrows -> BodyResult { ... } 
+}
+```
+
+A nursery can be launched from any asychronous context, eventually returns a single value (the `BodyResult`). Tasks many be added to it dynamically, as we saw in the `chopVegetables` example in the *Proposes solution: Nurseries* section, and the nursery enforces awaiting for all tasks before it returns by asserting that is is empty when returning the final result.
```
Comment (rjmccall) [inline proposals/nnnn-structured-concurrency.md]: Missing an "and" in the first sentence; there's a typo "Proposes"; "is is"

File: proposals/nnnn-structured-concurrency.md [hunk 4]
```diff
@@ -255,6 +257,148 @@ If the scope of an `async let` exits with a thrown error, the child task corresp
...
+    /// Query whether the nursery has any remaining tasks.
+    /// Nurseries are always empty upon entry to the withNursery body.
+    public var isEmpty: Bool { ... } 
+
+    /// Cancel all the remaining tasks in the nursery.
+    /// Any results, including errors thrown, are discarded.
+    public mutating func cancelAll() { ... } 
+  }
+}
+```
```
Comment (ktoso) [inline proposals/nnnn-structured-concurrency.md:397]: WIP

File: proposals/nnnn-structured-concurrency.md (hunk 1 above)
Comment (rjmccall) [reply proposals/nnnn-structured-concurrency.md:501]: Yes, I think having `add` fail after cancellation is probably reasonable.

Comment (rjmccall) [reply proposals/nnnn-structured-concurrency.md:501]: You're missing an example about "cancelled from the outside".  Are you talking about the task that created the nursery being cancellable?

File: proposals/nnnn-structured-concurrency.md [hunk 5]
```diff
@@ -255,6 +257,148 @@ If the scope of an `async let` exits with a thrown error, the child task corresp
...
+
+- first w add vegetable chopping tasks to the nursery
+- the chopping of the various vegetables beings asynchronously,
+- eventually an onion will be chopped and `throw`
+
+
+So far we did not yet discuss the cancellation of nurseries. A nursery can be cancelled form the outside, for example like this: 
+
+#### Nurseries: Implicitly awaited tasks
+Sometimes it is not necessary to gather the results of asynchronous functions (e.g. because they may be `Void` returning, "uni-directional"), in this case we can rely on the nursery implicitly awaiting for all tasks started before returning.
```
Comment (rjmccall) [inline proposals/nnnn-structured-concurrency.md:502]: Well, these aren't the semantics I described, but they might be more reasonable.

File: proposals/nnnn-structured-concurrency.md [hunk 6]
```diff
@@ -255,6 +257,148 @@ If the scope of an `async let` exits with a thrown error, the child task corresp
...
+```
+
+Let us break up the `chopOnionsAndCarrots()` function into multiple steps to fully understand its semantics:
+
+- first w add vegetable chopping tasks to the nursery
+- the chopping of the various vegetables beings asynchronously,
+- eventually an onion will be chopped and `throw`
+
+
+So far we did not yet discuss the cancellation of nurseries. A nursery can be cancelled form the outside, for example like this:
```
Comment (ktoso) [inline proposals/nnnn-structured-concurrency.md]: So we currently _don't_ have an API proposed to make a nursery and then launch stuff with it...

```swift
// "kind of" exists but we didn't really design it completely IMHO
let nursery = Nursery()

await try nursery.add {  // awaits only the "add"
 await task() 
}
await try nursery.add {  // awaits only the "add"
 await task() 
}

sleep(200)

nursery.cancel() // maybe some other thread
```

but also one could:

```swift
let handle = Task.runDetached { 
Task.withNursery { ... }
}

handle.cancel()
```

which is what I was going to show here for now.

File: proposals/nnnn-structured-concurrency.md (hunk 3 above)
Comment (ktoso) [reply proposals/nnnn-structured-concurrency.md]: Whoops thanks

File: proposals/nnnn-structured-concurrency.md (hunk 1 above)
Comment (ktoso) [reply proposals/nnnn-structured-concurrency.md:501]: Yes, today that's how it'd normally happen -- the task that started the nursery got cancelled so it cancels the nursery too.

Added mini example here: https://github.com/DougGregor/swift-evolution/pull/13/files#r513832435 before writing it up completely

File: proposals/nnnn-structured-concurrency.md (hunk 6 above)
Comment (rjmccall) [reply proposals/nnnn-structured-concurrency.md]: I don't think we'd have the first, I don't want arbitrary code doing stuff with nurseries.

File: proposals/nnnn-structured-concurrency.md (hunk 5 above)
Comment (ktoso) [reply proposals/nnnn-structured-concurrency.md:502]: I see okey... The Task nurseries section in our "big doc" did not really have much detail and that's how I understood the 

  // Postcondition: if the body returns normally, the nursery is empty.
  // If it throws, all tasks in the nursery will be automatically cancelled.

File: proposals/nnnn-structured-concurrency.md (hunk 6 above)
Comment (ktoso) [reply proposals/nnnn-structured-concurrency.md]: Okey got it, I wonder if that will really be doable but going to assume so for now then -- will only doc the parent task getting cancelled then 👍

File: proposals/nnnn-structured-concurrency.md (hunk 5 above)
Comment (ktoso) [reply proposals/nnnn-structured-concurrency.md:502]: I checked around a bit to confirm my thinking here, the same seems to be the semantics exhibited by  Trio as well as Kotlin scopes. I think it's a very useful semantic to have, for reference:

- https://trio.readthedocs.io/en/stable/tutorial.html#okay-let-s-see-something-cool-already
- https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-scope/index.html#custom-usage

> By convention, they all wait for all the coroutines inside their block to complete before completing themselves, thus enforcing the structured concurrency. See Job documentation for more details.

File: proposals/nnnn-structured-concurrency.md (hunk 6 above)
Comment (ktoso) [reply proposals/nnnn-structured-concurrency.md]: I documented the latter: https://github.com/DougGregor/swift-evolution/pull/13/commits/19625c4e65d3d3a66ea09663b796077ae7a61079 -- comment on the commit with thoughts...

**Pull request merged**